pkg/types/kustomization.go

```diff
@@ -376,3 +381,39 @@ type Replica struct {
 	// The number of replicas required.
 	Count int64 `json:"count,omitempty" yaml:"count,omitempty"`
 }
+
+// Patch represent either a Strategic Merge Patch or a JSON patch
+// and its targets.
+// The content of the patch can either be from a file
+// or from an inline string.
+type Patch struct {
+	// Path is a relative file path to the patch file.
+	Path string `json:"path,omitempty" yaml:"path,omitempty"`
+
+	// Patch is the content of a patch.
+	Patch string `json:"patch,omitempty" yaml:"patch,omitempty"`
+
+	// Target points to the resources that the patch is applied to
+	Target Target `json:"target,omitempty" yaml:"target,omitempty"`
+
+	// Type is one of `StrategicMergePatch` or `JsonPatch`.
+	// optional
+	Type string `json:"type,omitempty" yaml:"type,omitempty"`
+}
+
+// Targets specifies a set of resources.
+// Any resource that matches intersection of all conditions
+// is included in this set.
+type Targets struct {
+	gvk.Gvk   `json:",inline,omitempty" yaml:",inline,omitempty"`
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+	Name      string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// MatchAnnotations is a map of key-value pairs.
+	// It matches with the resource annotations.
+	MatchAnnotation map[string]string `json:"matchAnnotation,omitempty" yaml:"matchAnnotation,omitempty"`
```

Comment:
Do MatchAnnotation and LabelSelector work the same way?
if so, use the same naming convention.

The term LabelSelector sound like it's going to support label selection _expressions_.  
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api

If that's the plan, the type should just be a bare string, not a map.

if it's a map, then how are all these matches going to be treated?  OR or AND?
Is there a NOT?
are label and annotation matching OR'd together or AND'd together?
If labels or annotations are specified and the Name is specified, is it an error?

Reply:
Thanks for the link. 
I think both the label selection expressions (a single string) and a map[string]string work.
The map will use AND for all key,value pairs to select.
The
One looks as
```
labelSelector: environment=production,tier=frontend
```
And the other looks as
```
labelSelector:
    environment: production
   tier: frontend
```
I think the later one looks better. We can change the field name to `targetLabel` to make it not confusing.

> are label and annotation matching OR'd together or AND'd together?
It's an AND of all requirements: annotations, labels, gvk, name, namespace

> If labels or annotations are specified and the Name is specified, is it an error?
Not necessary. The name can be a regexp.

Reply:
IMO, use a string.  It's future proof.  You can introduce OR, NOT etc. later.  The map structure locks you into AND only (or OR only).   There's a reason why k8s does it that way.  Plus there's probably a  library somewhere that could be reused to do the parsing.

Field comments should mention where regexp is allows.

Reply:
maybe use this package? https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/labels.go

Reply:
OK, then I'll change it to a string

Reply:
@monopole @sethpollack  How is the name `MatchAnnotation`? What about `AnnotationSelector`?

Reply:
voting "Selector" for both, if they behave like the existing LabelSelector

Reply:
updated to `AnnotationSelector` and `LabelSelector`.